perun-core/src/main/java/cz/metacentrum/perun/core/impl/PerunLocksUtils.java

```diff
@@ -0,0 +1,141 @@
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.Set;
+import java.util.Random;
+import java.util.HashMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * Methods and structures for working with locks on objects and actions.
+ *
+ * Created by Michal Stava stavamichal@gmail.com
+ */
+public class PerunLocksUtils {
+
+	private final static Logger log = LoggerFactory.getLogger(PerunLocksUtils.class);
+
+	//This empty object is used just for purpose of saving and identifying all locks for separate transaction
+	public static Object uniqueKey = new Object();
+
+	//Maps for saving and working with specific locks
+	private static ConcurrentHashMap<Group, ReadWriteLock> groupsLocks = new ConcurrentHashMap<>();
+	private static ConcurrentHashMap<Group, ConcurrentHashMap<Member, Lock>> groupsMembersLocks = new ConcurrentHashMap<>();
+
+	/**
+	 * Create transaction locks for combination of group and member (from list of members)
+	 * and also bind them to the transaction (as resource by Object uniqueKey)
+	 *
+	 * @param group the group
+	 * @param members list of members
+	 * @throws InternalErrorException
+	 * @throws InterruptedException
+	 */
+	public static void lockGroupMembership(Group group, List<Member> members) throws InternalErrorException, InterruptedException {
+		if(group == null) throw new InternalErrorException("Group can't be null when creating lock for group and list of members.");
+		if(members == null || members.isEmpty()) throw new InternalErrorException("Members can't be null or empty when creating lock for group and list of members.");
+
+		//Sort list of members strictly by ids (there is compareTo method in perunBean using ids for comparing)
+		Collections.sort(members);
+
+		List<Lock> returnedLocks = new ArrayList<>();
+		//ReentrantReadWriteLock readwriteLock = new ReentrantReadWriteLock();
+
+		try {
+			//try to lock all needed locks there
+			ReadWriteLock groupReadWriteLock = groupsLocks.get(group);
+			if(groupReadWriteLock == null) {
+				groupsLocks.putIfAbsent(group, new ReentrantReadWriteLock(true));
+				groupReadWriteLock = groupsLocks.get(group);
+			}
+			groupReadWriteLock.readLock().lock();
+			returnedLocks.add(groupReadWriteLock.readLock());
+			for (Member member : members) {
+				//Get members lock map by group if exists or create a new one
+				ConcurrentHashMap<Member, Lock> newLockForMember = groupsMembersLocks.get(group);
+				if(newLockForMember == null) {
+					groupsMembersLocks.putIfAbsent(group, new ConcurrentHashMap<Member, Lock>());
+					newLockForMember = groupsMembersLocks.get(group);
+				}
+
+				//Get concrete member lock from members lock map or create a new one if not exists
+				Lock memberLock = groupsMembersLocks.get(group).get(member);
```

Comment:
Yes, you are right, I overlooked this. It is fixed now and also variable is renamed to be more suitable.